SMB Connection Fails - "This client is not compatible with the server."

In the version 3.5.0.89, the SMB connection does not work. Adding a new connection to a Fritz!Box 7490 with latest firmware fails.

The logs show, among others, these errors:

```
[10-24 17:40:56.811 15440:16132 E/SmbComNegotiateResponse]
Signatures are required but the server does not support them

[10-24 17:40:56.813 15440:16131 W/System.err]
jcifs.smb.SmbException: Failed to connect: 0.0.0.0<00>/192.168.77.1
...
[10-24 17:40:56.814 15440:16131 W/System.err]
Caused by: jcifs.util.transport.TransportException: jcifs.smb.SmbException: This client is not compatible with the server.

```

See full log: [logcat_10-24-2020_17-41-04.txt](https://github.com/TeamAmaze/AmazeFileManager/files/5433136/logcat_10-24-2020_17-41-04.txt)

I had to downgrade to an older version. **Version 3.1.2 RC1 (latest in the F-Droid Repo) works without this problem!**

--- a/app/src/main/java/com/amaze/filemanager/asynchronous/asynctasks/MoveFiles.java
+++ b/app/src/main/java/com/amaze/filemanager/asynchronous/asynctasks/MoveFiles.java
@@ -21,7 +21,6 @@
 package com.amaze.filemanager.asynchronous.asynctasks;
 
 import java.io.File;
-import java.net.MalformedURLException;
 import java.util.ArrayList;
 import java.util.HashSet;
 import java.util.List;
@@ -51,9 +50,6 @@ import android.content.Intent;
 import android.os.AsyncTask;
 import android.util.Log;
 import android.widget.Toast;
-
-import jcifs.smb.SmbException;
-import jcifs.smb.SmbFile;
 
 /**
  * AsyncTask that moves files from source to destination by trying to rename files first, if they're
@@ -97,6 +93,8 @@ public class MoveFiles extends AsyncTask<ArrayList<String>, String, Boolean> {
     for (int i = 0; i < paths.size(); i++) {
       for (HybridFileParcelable baseFile : files.get(i)) {
         String destPath = paths.get(i) + "/" + baseFile.getName(context);
+        if (baseFile.getPath().indexOf('?') > 0)
+          destPath += baseFile.getPath().substring(baseFile.getPath().indexOf('?'));
         if (!isMoveOperationValid(baseFile, new HybridFile(mode, paths.get(i)))) {
           // TODO: 30/06/20 Replace runtime exception with generic exception
           Log.w(
@@ -105,19 +103,6 @@ public class MoveFiles extends AsyncTask<ArrayList<String>, String, Boolean> {
           continue;
         }
         switch (mode) {
-          case SMB:
-            try {
-              SmbFile source = new SmbFile(baseFile.getPath());
-              SmbFile dest = new SmbFile(destPath);
-              source.renameTo(dest);
-            } catch (MalformedURLException e) {
-              e.printStackTrace();
-              return false;
-            } catch (SmbException e) {
-              e.printStackTrace();
-              return false;
-            }
-            break;
           case FILE:
             File dest = new File(destPath);
             File source = new File(baseFile.getPath());

--- a/app/src/main/java/com/amaze/filemanager/database/UtilsHandler.java
+++ b/app/src/main/java/com/amaze/filemanager/database/UtilsHandler.java
@@ -379,12 +379,18 @@ public class UtilsHandler {
   private void removeSmbPath(String name, String path) {
     if ("".equals(path))
       utilitiesDatabase.smbEntryDao().deleteByName(name).subscribeOn(Schedulers.io()).subscribe();
-    else
+    else {
+      try {
+        path = SmbUtil.getSmbEncryptedPath(context, path);
+      } catch (GeneralSecurityException | IOException e) {
+        Log.e(TAG, "Error encrypting path", e);
+      }
       utilitiesDatabase
           .smbEntryDao()
           .deleteByNameAndPath(name, path)
           .subscribeOn(Schedulers.io())
           .subscribe();
+    }
   }
 
   private void removeSftpPath(String name, String path) {

--- a/app/src/main/java/com/amaze/filemanager/filesystem/FileUtil.java
+++ b/app/src/main/java/com/amaze/filemanager/filesystem/FileUtil.java
@@ -32,7 +32,6 @@ import java.io.OutputStream;
 import java.io.OutputStreamWriter;
 import java.lang.reflect.InvocationTargetException;
 import java.lang.reflect.Method;
-import java.net.MalformedURLException;
 import java.nio.channels.FileChannel;
 import java.util.ArrayList;
 import java.util.List;
@@ -52,6 +51,7 @@ import com.amaze.filemanager.utils.AppConstants;
 import com.amaze.filemanager.utils.DataUtils;
 import com.amaze.filemanager.utils.OTGUtil;
 import com.amaze.filemanager.utils.OpenMode;
+import com.amaze.filemanager.utils.SmbUtil;
 import com.cloudrail.si.interfaces.CloudStorage;
 
 import android.annotation.TargetApi;
@@ -279,7 +279,7 @@ public abstract class FileUtil {
                           retval.add(targetFile.getPath());
                           break;
                         case SMB:
-                          SmbFile targetSmbFile = new SmbFile(finalFilePath);
+                          SmbFile targetSmbFile = SmbUtil.create(finalFilePath);
                           if (targetSmbFile.exists()) {
                             AppConfig.toast(
                                 mainActivity, mainActivity.getString(R.string.cannot_overwrite));
@@ -577,11 +577,8 @@ public abstract class FileUtil {
     switch (file.mode) {
       case SMB:
         try {
-          SmbFile smbFile = new SmbFile(file.getPath());
+          SmbFile smbFile = file.getSmbFile();
           smbFile.mkdirs();
-        } catch (MalformedURLException e) {
-          e.printStackTrace();
-          isSuccessful = false;
         } catch (SmbException e) {
           e.printStackTrace();
           isSuccessful = false;

--- a/app/src/main/java/com/amaze/filemanager/filesystem/FileUtil.java
+++ b/app/src/main/java/com/amaze/filemanager/filesystem/FileUtil.java
@@ -20,6 +20,9 @@
 
 package com.amaze.filemanager.filesystem;
 
+import static com.amaze.filemanager.filesystem.smb.CifsContextFactory.SMB_URI_PREFIX;
+import static com.amaze.filemanager.filesystem.ssh.SshConnectionPool.SSH_URI_PREFIX;
+
 import java.io.BufferedInputStream;
 import java.io.BufferedOutputStream;
 import java.io.File;
@@ -287,7 +290,8 @@ public abstract class FileUtil {
                           } else {
                             OutputStream outputStream = targetSmbFile.getOutputStream();
                             bufferedOutputStream = new BufferedOutputStream(outputStream);
-                            retval.add(HybridFile.parseSmbPath(targetSmbFile.getPath()));
+                            retval.add(
+                                HybridFile.parseAndFormatUriForDisplay(targetSmbFile.getPath()));
                           }
                           break;
                         case SFTP:
@@ -1016,8 +1020,8 @@ public abstract class FileUtil {
    */
   public static int checkFolder(final String f, Context context) {
     if (f == null) return 0;
-    if (f.startsWith("smb://")
-        || f.startsWith("ssh://")
+    if (f.startsWith(SMB_URI_PREFIX)
+        || f.startsWith(SSH_URI_PREFIX)
         || f.startsWith(OTGUtil.PREFIX_OTG)
         || f.startsWith(CloudHandler.CLOUD_PREFIX_BOX)
         || f.startsWith(CloudHandler.CLOUD_PREFIX_GOOGLE_DRIVE)

--- a/app/src/main/java/com/amaze/filemanager/filesystem/HybridFile.java
+++ b/app/src/main/java/com/amaze/filemanager/filesystem/HybridFile.java
@@ -48,11 +48,13 @@ import com.amaze.filemanager.utils.DataUtils;
 import com.amaze.filemanager.utils.OTGUtil;
 import com.amaze.filemanager.utils.OnFileFound;
 import com.amaze.filemanager.utils.OpenMode;
+import com.amaze.filemanager.utils.SmbUtil;
 import com.cloudrail.si.interfaces.CloudStorage;
 import com.cloudrail.si.types.SpaceAllocation;
 
 import android.content.ContentResolver;
 import android.content.Context;
+import android.net.Uri;
 import android.os.Build;
 import android.preference.PreferenceManager;
 import android.util.Log;
@@ -93,9 +95,9 @@ public class HybridFile {
   public HybridFile(OpenMode mode, String path, String name, boolean isDirectory) {
     this(mode, path);
     if (path.startsWith("smb://") || isSmb()) {
-      if (!isDirectory) this.path += name;
-      else if (!name.endsWith("/")) this.path += name + "/";
-      else this.path += name;
+      Uri.Builder pathBuilder = Uri.parse(this.path).buildUpon().appendPath(name);
+      if (isDirectory) pathBuilder.appendPath("/");
+      this.path = pathBuilder.build().toString();
     } else if (path.startsWith("ssh://") || isSftp()) {
       this.path += "/" + name;
     } else if (isRoot() && path.equals("/")) {
@@ -329,7 +331,7 @@ public class HybridFile {
 
   public SmbFile getSmbFile(int timeout) {
     try {
-      SmbFile smbFile = new SmbFile(path);
+      SmbFile smbFile = SmbUtil.create(path);
       smbFile.setConnectTimeout(timeout);
       return smbFile;
     } catch (MalformedURLException e) {
@@ -339,7 +341,7 @@ public class HybridFile {
 
   public SmbFile getSmbFile() {
     try {
-      return new SmbFile(path);
+      return SmbUtil.create(path);
     } catch (MalformedURLException e) {
       return null;
     }
@@ -361,12 +363,8 @@ public class HybridFile {
     String parentPath = "";
     switch (mode) {
       case SMB:
-        try {
-          parentPath = new SmbFile(path).getParent();
-        } catch (MalformedURLException e) {
-          parentPath = "";
-          e.printStackTrace();
-        }
+        SmbFile smbFile = getSmbFile();
+        parentPath = (smbFile != null) ? smbFile.getParent() : "";
         break;
       case FILE:
       case ROOT:
@@ -406,14 +404,11 @@ public class HybridFile {
       case SFTP:
         return isDirectory(AppConfig.getInstance());
       case SMB:
+        SmbFile smbFile = getSmbFile();
         try {
-          isDirectory = new SmbFile(path).isDirectory();
+          isDirectory = smbFile != null && smbFile.isDirectory();
         } catch (SmbException e) {
           isDirectory = false;
-          e.printStackTrace();
-        } catch (MalformedURLException e) {
-          isDirectory = false;
-          e.printStackTrace();
         }
         break;
       case FILE:
@@ -464,7 +459,7 @@ public class HybridFile {
       case SMB:
         try {
           isDirectory =
-              Single.fromCallable(() -> new SmbFile(path).isDirectory())
+              Single.fromCallable(() -> getSmbFile().isDirectory())
                   .subscribeOn(Schedulers.io())
                   .blockingGet();
         } catch (Exception e) {
@@ -530,12 +525,8 @@ public class HybridFile {
       case SFTP:
         return folderSize(AppConfig.getInstance());
       case SMB:
-        try {
-          size = FileUtils.folderSize(new SmbFile(path));
-        } catch (MalformedURLException e) {
-          size = 0L;
-          e.printStackTrace();
-        }
+        SmbFile smbFile = getSmbFile();
+        size = smbFile != null ? FileUtils.folderSize(getSmbFile()) : 0;
         break;
       case FILE:
         size = FileUtils.folderSize(getFile(), null);
@@ -565,12 +556,8 @@ public class HybridFile {
               }
             });
       case SMB:
-        try {
-          size = FileUtils.folderSize(new SmbFile(path));
-        } catch (MalformedURLException e) {
-          size = 0l;
-          e.printStackTrace();
-        }
+        SmbFile smbFile = getSmbFile();
+        size = (smbFile != null) ? FileUtils.folderSize(smbFile) : 0L;
         break;
       case FILE:
         size = FileUtils.folderSize(getFile(), null);
@@ -602,10 +589,8 @@ public class HybridFile {
     switch (mode) {
       case SMB:
         try {
-          size = (new SmbFile(path).getDiskFreeSpace());
-        } catch (MalformedURLException e) {
-          size = 0L;
-          e.printStackTrace();
+          SmbFile smbFile = getSmbFile();
+          size = smbFile != null ? smbFile.getDiskFreeSpace() : 0L;
         } catch (SmbException e) {
           size = 0L;
           e.printStackTrace();
@@ -663,10 +648,9 @@ public class HybridFile {
       case SMB:
         // TODO: Find total storage space of SMB when JCIFS adds support
         try {
-          size = new SmbFile(path).getDiskFreeSpace();
+          SmbFile smbFile = getSmbFile();
+          size = smbFile != null ? smbFile.getDiskFreeSpace() : 0L;
         } catch (SmbException e) {
-          e.printStackTrace();
-        } catch (MalformedURLException e) {
           e.printStackTrace();
         }
         break;
@@ -751,12 +735,21 @@ public class HybridFile {
         break;
       case SMB:
         try {
-          SmbFile smbFile = new SmbFile(path);
-          for (SmbFile smbFile1 : smbFile.listFiles()) {
-            HybridFileParcelable baseFile = new HybridFileParcelable(smbFile1);
-            onFileFound.onFileFound(baseFile);
+          SmbFile smbFile = getSmbFile();
+          if (smbFile != null) {
+            for (SmbFile smbFile1 : smbFile.listFiles()) {
+              HybridFileParcelable baseFile;
+              try {
+                SmbFile sf = new SmbFile(smbFile1.getURL(), smbFile.getContext());
+                baseFile = new HybridFileParcelable(sf);
+              } catch (MalformedURLException shouldNeverHappen) {
+                shouldNeverHappen.printStackTrace();
+                baseFile = new HybridFileParcelable(smbFile1);
+              }
+              onFileFound.onFileFound(baseFile);
+            }
           }
-        } catch (MalformedURLException | SmbException e) {
+        } catch (SmbException e) {
           e.printStackTrace();
         }
         break;
@@ -830,14 +823,13 @@ public class HybridFile {
         break;
       case SMB:
         try {
-          SmbFile smbFile = new SmbFile(path);
-          for (SmbFile smbFile1 : smbFile.listFiles()) {
-            HybridFileParcelable baseFile = new HybridFileParcelable(smbFile1);
-            arrayList.add(baseFile);
+          SmbFile smbFile = getSmbFile();
+          if (smbFile != null) {
+            for (SmbFile smbFile1 : smbFile.listFiles()) {
+              HybridFileParcelable baseFile = new HybridFileParcelable(smbFile1);
+              arrayList.add(baseFile);
+            }
           }
-        } catch (MalformedURLException e) {
-          arrayList.clear();
-          e.printStackTrace();
         } catch (SmbException e) {
           arrayList.clear();
           e.printStackTrace();
@@ -865,19 +857,17 @@ public class HybridFile {
   }
 
   public String getReadablePath(String path) {
-    if (isSftp()) return parseSftpPath(path);
-    if (isSmb()) return parseSmbPath(path);
-    return path;
+    if (isSftp() || isSmb()) return parseSmbPath(path);
+    else return path;
   }
 
   public static String parseSftpPath(String a) {
-    if (a.contains("@")) return "ssh://" + a.substring(a.indexOf("@") + 1, a.length());
-    else return a;
+    return parseSmbPath(a);
   }
 
   public static String parseSmbPath(String a) {
-    if (a.contains("@")) return "smb://" + a.substring(a.indexOf("@") + 1, a.length());
-    else return a;
+    Uri uri = Uri.parse(a);
+    return String.format("%s://%s%s", uri.getScheme(), uri.getHost(), uri.getPath());
   }
 
   /**
@@ -907,7 +897,7 @@ public class HybridFile {
           });
     } else if (isSmb()) {
       try {
-        inputStream = new SmbFile(path).getInputStream();
+        inputStream = getSmbFile().getInputStream();
       } catch (IOException e) {
         inputStream = null;
         e.printStackTrace();
@@ -950,7 +940,7 @@ public class HybridFile {
         break;
       case SMB:
         try {
-          inputStream = new SmbFile(path).getInputStream();
+          inputStream = getSmbFile().getInputStream();
         } catch (IOException e) {
           inputStream = null;
           e.printStackTrace();
@@ -1029,7 +1019,7 @@ public class HybridFile {
             });
       case SMB:
         try {
-          outputStream = new SmbFile(path).getOutputStream();
+          outputStream = getSmbFile().getOutputStream();
         } catch (IOException e) {
           outputStream = null;
           e.printStackTrace();
@@ -1128,11 +1118,14 @@ public class HybridFile {
   public boolean setLastModified(final long date) {
     if (isSmb()) {
       try {
-        new SmbFile(path).setLastModified(date);
-        return true;
+        SmbFile smbFile = getSmbFile();
+        if (smbFile != null) {
+          smbFile.setLastModified(date);
+          return true;
+        } else {
+          return false;
+        }
       } catch (SmbException e) {
-        return false;
-      } catch (MalformedURLException e) {
         return false;
       }
     }
@@ -1156,8 +1149,8 @@ public class HybridFile {
           });
     } else if (isSmb()) {
       try {
-        new SmbFile(path).mkdirs();
-      } catch (SmbException | MalformedURLException e) {
+        getSmbFile().mkdirs();
+      } catch (SmbException e) {
         e.printStackTrace();
       }
     } else if (isOtgFile()) {
@@ -1213,8 +1206,8 @@ public class HybridFile {
       return true;
     } else if (isSmb()) {
       try {
-        new SmbFile(path).delete();
-      } catch (SmbException | MalformedURLException e) {
+        getSmbFile().delete();
+      } catch (SmbException e) {
         e.printStackTrace();
       }
     } else {

--- a/app/src/main/java/com/amaze/filemanager/filesystem/HybridFile.java
+++ b/app/src/main/java/com/amaze/filemanager/filesystem/HybridFile.java
@@ -20,6 +20,9 @@
 
 package com.amaze.filemanager.filesystem;
 
+import static com.amaze.filemanager.filesystem.smb.CifsContextFactory.SMB_URI_PREFIX;
+import static com.amaze.filemanager.filesystem.ssh.SshConnectionPool.SSH_URI_PREFIX;
+
 import java.io.File;
 import java.io.FileInputStream;
 import java.io.FileNotFoundException;
@@ -94,11 +97,11 @@ public class HybridFile {
 
   public HybridFile(OpenMode mode, String path, String name, boolean isDirectory) {
     this(mode, path);
-    if (path.startsWith("smb://") || isSmb()) {
+    if (path.startsWith(SMB_URI_PREFIX) || isSmb()) {
       Uri.Builder pathBuilder = Uri.parse(this.path).buildUpon().appendPath(name);
       if (isDirectory) pathBuilder.appendPath("/");
       this.path = pathBuilder.build().toString();
-    } else if (path.startsWith("ssh://") || isSftp()) {
+    } else if (path.startsWith(SSH_URI_PREFIX) || isSftp()) {
       this.path += "/" + name;
     } else if (isRoot() && path.equals("/")) {
       // root of filesystem, don't concat another '/'
@@ -109,9 +112,9 @@ public class HybridFile {
   }
 
   public void generateMode(Context context) {
-    if (path.startsWith("smb://")) {
+    if (path.startsWith(SMB_URI_PREFIX)) {
       mode = OpenMode.SMB;
-    } else if (path.startsWith("ssh://")) {
+    } else if (path.startsWith(SSH_URI_PREFIX)) {
       mode = OpenMode.SFTP;
     } else if (path.startsWith(OTGUtil.PREFIX_OTG)) {
       mode = OpenMode.OTG;
@@ -408,6 +411,7 @@ public class HybridFile {
         try {
           isDirectory = smbFile != null && smbFile.isDirectory();
         } catch (SmbException e) {
+          e.printStackTrace();
           isDirectory = false;
         }
         break;
@@ -857,16 +861,12 @@ public class HybridFile {
   }
 
   public String getReadablePath(String path) {
-    if (isSftp() || isSmb()) return parseSmbPath(path);
+    if (isSftp() || isSmb()) return parseAndFormatUriForDisplay(path);
     else return path;
   }
 
-  public static String parseSftpPath(String a) {
-    return parseSmbPath(a);
-  }
-
-  public static String parseSmbPath(String a) {
-    Uri uri = Uri.parse(a);
+  public static String parseAndFormatUriForDisplay(@NonNull String uriString) {
+    Uri uri = Uri.parse(uriString);
     return String.format("%s://%s%s", uri.getScheme(), uri.getHost(), uri.getPath());
   }
 

--- a/app/src/main/java/com/amaze/filemanager/filesystem/Operations.java
+++ b/app/src/main/java/com/amaze/filemanager/filesystem/Operations.java
@@ -25,6 +25,7 @@ import java.io.File;
 import java.io.IOException;
 import java.io.OutputStream;
 import java.net.MalformedURLException;
+import java.net.URL;
 
 import com.amaze.filemanager.exceptions.ShellNotRunningException;
 import com.amaze.filemanager.filesystem.cloud.CloudUtil;
@@ -386,17 +387,15 @@ public class Operations {
 
         if (oldFile.isSmb()) {
           try {
-            SmbFile smbFile = new SmbFile(oldFile.getPath());
-            SmbFile smbFile1 = new SmbFile(newFile.getPath());
+            SmbFile smbFile = oldFile.getSmbFile();
+            SmbFile smbFile1 = new SmbFile(new URL(newFile.getPath()), smbFile.getContext());
             if (smbFile1.exists()) {
               errorCallBack.exists(newFile);
               return null;
             }
             smbFile.renameTo(smbFile1);
             if (!smbFile.exists() && smbFile1.exists()) errorCallBack.done(newFile, true);
-          } catch (MalformedURLException e) {
-            e.printStackTrace();
-          } catch (SmbException e) {
+          } catch (SmbException | MalformedURLException e) {
             e.printStackTrace();
           }
           return null;

--- a/app/src/main/java/com/amaze/filemanager/filesystem/cloud/CloudUtil.java
+++ b/app/src/main/java/com/amaze/filemanager/filesystem/cloud/CloudUtil.java
@@ -58,7 +58,6 @@ import android.widget.Toast;
 
 import androidx.documentfile.provider.DocumentFile;
 
-import jcifs.smb.SmbFile;
 import net.schmizz.sshj.sftp.RemoteFile;
 import net.schmizz.sshj.sftp.SFTPClient;
 
@@ -320,7 +319,7 @@ public class CloudUtil {
         break;
       case SMB:
         try {
-          inputStream = new SmbFile(hybridFile.getPath()).getInputStream();
+          inputStream = hybridFile.getSmbFile().getInputStream();
         } catch (IOException e) {
           inputStream = null;
           e.printStackTrace();

--- a/app/src/main/java/com/amaze/filemanager/filesystem/smb/CifsContextFactory.java
+++ b/app/src/main/java/com/amaze/filemanager/filesystem/smb/CifsContextFactory.java
@@ -36,6 +36,8 @@ import jcifs.context.SingletonContext;
 
 public abstract class CifsContextFactory {
 
+  public static final String SMB_URI_PREFIX = "smb://";
+
   private static final String TAG = CifsContextFactory.class.getSimpleName();
 
   private static final Properties defaultProperties;

--- a/app/src/main/java/com/amaze/filemanager/filesystem/smb/CifsContextFactory.java
+++ b/app/src/main/java/com/amaze/filemanager/filesystem/smb/CifsContextFactory.java
@@ -1,0 +1,77 @@
+/*
+ * Copyright (C) 2014-2020 Arpit Khurana <arpitkh96@gmail.com>, Vishal Nehra <vishalmeham2@gmail.com>,
+ * Emmanuel Messulam<emmanuelbendavid@gmail.com>, Raymond Lai <airwave209gt at gmail.com> and Contributors.
+ *
+ * This file is part of Amaze File Manager.
+ *
+ * Amaze File Manager is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package com.amaze.filemanager.filesystem.smb;
+
+import java.util.Properties;
+
+import android.util.Log;
+
+import androidx.annotation.NonNull;
+import androidx.annotation.Nullable;
+
+import io.reactivex.Single;
+import io.reactivex.schedulers.Schedulers;
+import jcifs.CIFSException;
+import jcifs.config.PropertyConfiguration;
+import jcifs.context.BaseContext;
+import jcifs.context.SingletonContext;
+
+public abstract class CifsContextFactory {
+
+  private static final String TAG = CifsContextFactory.class.getSimpleName();
+
+  private static final Properties defaultProperties;
+
+  static {
+    defaultProperties = new Properties();
+    defaultProperties.setProperty("jcifs.resolveOrder", "BCAST");
+    defaultProperties.setProperty("jcifs.smb.client.responseTimeout", "30000");
+    defaultProperties.setProperty("jcifs.netbios.retryTimeout", "5000");
+    defaultProperties.setProperty("jcifs.netbios.cachePolicy", "-1");
+  }
+
+  public static final @NonNull BaseContext createWithDisableIpcSigningCheck(
+      boolean disableIpcSigningCheck) {
+    if (disableIpcSigningCheck) {
+      Properties extraProperties = new Properties();
+      extraProperties.put("jcifs.smb.client.ipcSigningEnforced", "false");
+      return create(extraProperties);
+    } else {
+      return create(null);
+    }
+  }
+
+  public static final @NonNull BaseContext create(@Nullable final Properties extraProperties) {
+    return Single.fromCallable(
+            () -> {
+              try {
+                Properties p = new Properties(defaultProperties);
+                if (extraProperties != null) p.putAll(extraProperties);
+                return new BaseContext(new PropertyConfiguration(p));
+              } catch (CIFSException e) {
+                Log.e(TAG, "Error initialize jcifs BaseContext, returning default", e);
+                return SingletonContext.getInstance();
+              }
+            })
+        .subscribeOn(Schedulers.io())
+        .blockingGet();
+  }
+}

--- a/app/src/main/java/com/amaze/filemanager/ui/dialogs/RenameBookmark.java
+++ b/app/src/main/java/com/amaze/filemanager/ui/dialogs/RenameBookmark.java
@@ -20,6 +20,8 @@
 
 package com.amaze.filemanager.ui.dialogs;
 
+import static com.amaze.filemanager.filesystem.smb.CifsContextFactory.SMB_URI_PREFIX;
+
 import java.net.URL;
 import java.net.URLDecoder;
 import java.net.URLEncoder;
@@ -112,7 +114,7 @@ public class RenameBookmark extends DialogFragment {
           });
       final AppCompatEditText ip = v2.findViewById(R.id.editText);
       if (studiomode != 0) {
-        if (path.startsWith("smb:/")) {
+        if (path.startsWith(SMB_URI_PREFIX)) {
           try {
             URL a = new URL(path);
             String userinfo = a.getUserInfo();
@@ -121,7 +123,7 @@ public class RenameBookmark extends DialogFragment {
               user = inf.substring(0, inf.indexOf(":"));
               pass = inf.substring(inf.indexOf(":") + 1, inf.length());
               String ipp = a.getHost();
-              pa = "smb://" + ipp + a.getPath();
+              pa = SMB_URI_PREFIX + ipp + a.getPath();
             }
           } catch (Exception e) {
             e.printStackTrace();
@@ -146,13 +148,13 @@ public class RenameBookmark extends DialogFragment {
               v -> {
                 String t = ip.getText().toString();
                 String name = conName.getText().toString();
-                if (studiomode != 0 && t.startsWith("smb://")) {
+                if (studiomode != 0 && t.startsWith(SMB_URI_PREFIX)) {
                   try {
                     URL a = new URL(t);
                     String userinfo = a.getUserInfo();
                     if (userinfo == null && user.length() > 0) {
                       t =
-                          "smb://"
+                          SMB_URI_PREFIX
                               + ((URLEncoder.encode(user, "UTF-8")
                                   + ":"
                                   + URLEncoder.encode(pass, "UTF-8")

--- a/app/src/main/java/com/amaze/filemanager/ui/dialogs/SmbConnectDialog.java
+++ b/app/src/main/java/com/amaze/filemanager/ui/dialogs/SmbConnectDialog.java
@@ -20,6 +20,8 @@
 
 package com.amaze.filemanager.ui.dialogs;
 
+import static com.amaze.filemanager.utils.SmbUtil.PARAM_DISABLE_IPC_SIGNING_CHECK;
+
 import java.io.IOException;
 import java.io.UnsupportedEncodingException;
 import java.net.MalformedURLException;
@@ -30,6 +32,7 @@ import java.security.GeneralSecurityException;
 
 import com.afollestad.materialdialogs.MaterialDialog;
 import com.amaze.filemanager.R;
+import com.amaze.filemanager.filesystem.smb.CifsContextFactory;
 import com.amaze.filemanager.ui.activities.superclasses.BasicActivity;
 import com.amaze.filemanager.ui.activities.superclasses.ThemedActivity;
 import com.amaze.filemanager.ui.provider.UtilitiesProvider;
@@ -45,6 +48,7 @@ import android.content.Context;
 import android.content.SharedPreferences;
 import android.content.res.ColorStateList;
 import android.graphics.Color;
+import android.net.UrlQuerySanitizer;
 import android.os.Build;
 import android.os.Bundle;
 import android.preference.PreferenceManager;
@@ -178,23 +182,25 @@ public class SmbConnectDialog extends DialogFragment {
 
     int accentColor = ((ThemedActivity) getActivity()).getAccent();
     final AppCompatEditText pass = v2.findViewById(R.id.passwordET);
-    final AppCompatCheckBox ch = v2.findViewById(R.id.checkBox2);
+    final AppCompatCheckBox chkSmbAnonymous = v2.findViewById(R.id.chkSmbAnonymous);
+    final AppCompatCheckBox chkSmbDisableIpcSignature =
+        v2.findViewById(R.id.chkSmbDisableIpcSignature);
     TextView help = v2.findViewById(R.id.wanthelp);
 
     EditTextColorStateUtil.setTint(context, conName, accentColor);
     EditTextColorStateUtil.setTint(context, user, accentColor);
     EditTextColorStateUtil.setTint(context, pass, accentColor);
 
-    Utils.setTint(context, ch, accentColor);
+    Utils.setTint(context, chkSmbAnonymous, accentColor);
     help.setOnClickListener(
         v -> {
           int accentColor1 = ((ThemedActivity) getActivity()).getAccent();
           GeneralDialogCreation.showSMBHelpDialog(context, accentColor1);
         });
 
-    ch.setOnClickListener(
+    chkSmbAnonymous.setOnClickListener(
         view -> {
-          if (ch.isChecked()) {
+          if (chkSmbAnonymous.isChecked()) {
             user.setEnabled(false);
             pass.setEnabled(false);
           } else {
@@ -219,9 +225,18 @@ public class SmbConnectDialog extends DialogFragment {
           domain.setText(domainp);
           user.setText(userp);
           pass.setText(passp);
-        } else ch.setChecked(true);
+        } else {
+          chkSmbAnonymous.setChecked(true);
+        }
         ipp = a.getHost();
         ip.setText(ipp);
+
+        UrlQuerySanitizer sanitizer = new UrlQuerySanitizer(path);
+        if (sanitizer.hasParameter(PARAM_DISABLE_IPC_SIGNING_CHECK)) {
+          chkSmbDisableIpcSignature.setChecked(
+              Boolean.parseBoolean(sanitizer.getValue(PARAM_DISABLE_IPC_SIGNING_CHECK)));
+        }
+
       } catch (UnsupportedEncodingException e) {
         e.printStackTrace();
       } catch (MalformedURLException e) {
@@ -273,15 +288,20 @@ public class SmbConnectDialog extends DialogFragment {
           }
           SmbFile smbFile;
           String domaind = domain.getText().toString();
-          if (ch.isChecked()) smbFile = createSMBPath(new String[] {ipa, "", "", domaind}, true);
+          if (chkSmbAnonymous.isChecked())
+            smbFile = createSMBPath(new String[] {ipa, "", "", domaind}, true, false);
           else {
             String useraw = user.getText().toString();
             String useru = useraw.replaceAll(" ", "\\ ");
             String passp = pass.getText().toString();
-            smbFile = createSMBPath(new String[] {ipa, useru, passp, domaind}, false);
+            smbFile = createSMBPath(new String[] {ipa, useru, passp, domaind}, false, false);
           }
 
           if (smbFile == null) return;
+
+          StringBuilder extraParams = new StringBuilder();
+          if (chkSmbDisableIpcSignature.isChecked())
+            extraParams.append(PARAM_DISABLE_IPC_SIGNING_CHECK).append('=').append(true);
 
           try {
             s =
@@ -297,7 +317,9 @@ public class SmbConnectDialog extends DialogFragment {
 
           if (smbConnectionListener != null) {
             // encrypted path means path with encrypted pass
-            smbConnectionListener.addConnection(edit, s[0], smbFile.getPath(), s[1], name, path);
+            String qs = extraParams.length() > 0 ? extraParams.insert(0, '?').toString() : "";
+            smbConnectionListener.addConnection(
+                edit, s[0], smbFile.getPath() + qs, s[1] + qs, name, path);
           }
           dismiss();
         });
@@ -314,24 +336,23 @@ public class SmbConnectDialog extends DialogFragment {
     return ba3.build();
   }
 
-  private SmbFile createSMBPath(String[] auth, boolean anonym) {
+  private SmbFile createSMBPath(String[] auth, boolean anonymous, boolean disableIpcSignCheck) {
     try {
       String yourPeerIP = auth[0], domain = auth[3];
 
-      String path =
-          "smb://"
-              + (android.text.TextUtils.isEmpty(domain)
-                  ? ""
-                  : (URLEncoder.encode(domain + ";", "UTF-8")))
-              + (anonym
-                  ? ""
-                  : (URLEncoder.encode(auth[1], "UTF-8")
-                      + ":"
-                      + URLEncoder.encode(auth[2], "UTF-8")
-                      + "@"))
-              + yourPeerIP
-              + "/";
-      SmbFile smbFile = new SmbFile(path);
+      StringBuilder sb = new StringBuilder("smb://");
+      if (!android.text.TextUtils.isEmpty(domain))
+        sb.append(URLEncoder.encode(domain + ";", "UTF-8"));
+      if (!anonymous)
+        sb.append(URLEncoder.encode(auth[1], "UTF-8"))
+            .append(":")
+            .append(URLEncoder.encode(auth[2], "UTF-8"))
+            .append("@");
+      sb.append(yourPeerIP).append("/");
+      SmbFile smbFile =
+          new SmbFile(
+              sb.toString(),
+              CifsContextFactory.createWithDisableIpcSigningCheck(disableIpcSignCheck));
       return smbFile;
     } catch (MalformedURLException e) {
       e.printStackTrace();

--- a/app/src/main/java/com/amaze/filemanager/ui/dialogs/SmbConnectDialog.java
+++ b/app/src/main/java/com/amaze/filemanager/ui/dialogs/SmbConnectDialog.java
@@ -20,6 +20,7 @@
 
 package com.amaze.filemanager.ui.dialogs;
 
+import static com.amaze.filemanager.filesystem.smb.CifsContextFactory.SMB_URI_PREFIX;
 import static com.amaze.filemanager.utils.SmbUtil.PARAM_DISABLE_IPC_SIGNING_CHECK;
 
 import java.io.IOException;
@@ -340,7 +341,7 @@ public class SmbConnectDialog extends DialogFragment {
     try {
       String yourPeerIP = auth[0], domain = auth[3];
 
-      StringBuilder sb = new StringBuilder("smb://");
+      StringBuilder sb = new StringBuilder(SMB_URI_PREFIX);
       if (!android.text.TextUtils.isEmpty(domain))
         sb.append(URLEncoder.encode(domain + ";", "UTF-8"));
       if (!anonymous)

--- a/app/src/main/java/com/amaze/filemanager/ui/fragments/MainFragment.java
+++ b/app/src/main/java/com/amaze/filemanager/ui/fragments/MainFragment.java
@@ -20,6 +20,7 @@
 
 package com.amaze.filemanager.ui.fragments;
 
+import static com.amaze.filemanager.filesystem.ssh.SshConnectionPool.SSH_URI_PREFIX;
 import static com.amaze.filemanager.ui.fragments.preference_fragments.PreferencesConstants.PREFERENCE_DIRECTORY_SORT_MODE;
 import static com.amaze.filemanager.ui.fragments.preference_fragments.PreferencesConstants.PREFERENCE_GRID_COLUMNS;
 import static com.amaze.filemanager.ui.fragments.preference_fragments.PreferencesConstants.PREFERENCE_SHOW_DIVIDERS;
@@ -1249,7 +1250,7 @@ public class MainFragment extends Fragment implements BottomBarButtonPath {
               loadlist(path.toString().replace("%3D", "="), true, openMode);
             } else loadlist(home, false, OpenMode.FILE);
           } else if (OpenMode.SFTP.equals(openMode)) {
-            if (!CURRENT_PATH.substring("ssh://".length()).contains("/"))
+            if (!CURRENT_PATH.substring(SSH_URI_PREFIX.length()).contains("/"))
               loadlist(home, false, OpenMode.FILE);
             else loadlist(currentFile.getParent(getContext()), true, openMode);
           } else if (CURRENT_PATH.equals("/")

--- a/app/src/main/java/com/amaze/filemanager/ui/fragments/MainFragment.java
+++ b/app/src/main/java/com/amaze/filemanager/ui/fragments/MainFragment.java
@@ -28,7 +28,6 @@ import static com.amaze.filemanager.ui.fragments.preference_fragments.Preference
 import static com.amaze.filemanager.ui.fragments.preference_fragments.PreferencesConstants.PREFERENCE_SHOW_THUMB;
 
 import java.io.File;
-import java.net.MalformedURLException;
 import java.util.ArrayList;
 import java.util.Collections;
 import java.util.HashMap;
@@ -92,6 +91,7 @@ import android.os.AsyncTask;
 import android.os.Build;
 import android.os.Bundle;
 import android.preference.PreferenceManager;
+import android.text.TextUtils;
 import android.text.format.Formatter;
 import android.util.Log;
 import android.view.LayoutInflater;
@@ -106,6 +106,7 @@ import android.widget.ImageView;
 import android.widget.TextView;
 import android.widget.Toast;
 
+import androidx.annotation.NonNull;
 import androidx.appcompat.view.ActionMode;
 import androidx.core.content.pm.ShortcutInfoCompat;
 import androidx.core.content.pm.ShortcutManagerCompat;
@@ -1185,7 +1186,7 @@ public class MainFragment extends Fragment implements BottomBarButtonPath {
                   .rename(
                       openMode,
                       f.getPath(),
-                      CURRENT_PATH + "/" + name1,
+                      Uri.parse(CURRENT_PATH).buildUpon().appendPath(name1).build().toString(),
                       getActivity(),
                       getMainActivity().isRootExplorer());
             },
@@ -1240,18 +1241,14 @@ public class MainFragment extends Fragment implements BottomBarButtonPath {
         if (selection) {
           adapter.toggleChecked(false);
         } else {
-
-          if (openMode == OpenMode.SMB) {
-            try {
-              if (!smbPath.equals(CURRENT_PATH)) {
-                String path = (new SmbFile(CURRENT_PATH).getParent());
-                loadlist((path), true, openMode);
-              } else loadlist(home, false, OpenMode.FILE);
-            } catch (MalformedURLException e) {
-              e.printStackTrace();
-            }
-
-          } else if (openMode == OpenMode.SFTP) {
+          if (OpenMode.SMB.equals(openMode)) {
+            if (!smbPath.equals(CURRENT_PATH)) {
+              StringBuilder path = new StringBuilder(currentFile.getSmbFile().getParent());
+              if (CURRENT_PATH.indexOf('?') > 0)
+                path.append(CURRENT_PATH.substring(CURRENT_PATH.indexOf('?')));
+              loadlist(path.toString().replace("%3D", "="), true, openMode);
+            } else loadlist(home, false, OpenMode.FILE);
+          } else if (OpenMode.SFTP.equals(openMode)) {
             if (!CURRENT_PATH.substring("ssh://".length()).contains("/"))
               loadlist(home, false, OpenMode.FILE);
             else loadlist(currentFile.getParent(getContext()), true, openMode);
@@ -1339,14 +1336,12 @@ public class MainFragment extends Fragment implements BottomBarButtonPath {
         adapter.toggleChecked(false);
       } else {
         if (openMode == OpenMode.SMB) {
-          try {
-            if (!CURRENT_PATH.equals(smbPath)) {
-              String path = (new SmbFile(CURRENT_PATH).getParent());
-              loadlist((path), true, OpenMode.SMB);
-            } else loadlist(home, false, OpenMode.FILE);
-          } catch (MalformedURLException e) {
-            e.printStackTrace();
-          }
+          if (!CURRENT_PATH.equals(smbPath)) {
+            StringBuilder path = new StringBuilder(currentFile.getSmbFile().getParent());
+            if (CURRENT_PATH.indexOf('?') > 0)
+              path.append(CURRENT_PATH.substring(CURRENT_PATH.indexOf('?')));
+            loadlist(path.toString(), true, OpenMode.SMB);
+          } else loadlist(home, false, OpenMode.FILE);
         } else if (CURRENT_PATH.equals("/")
             || CURRENT_PATH.equals(OTGUtil.PREFIX_OTG)
             || CURRENT_PATH.equals(CloudHandler.CLOUD_PREFIX_BOX + "/")
@@ -1412,8 +1407,10 @@ public class MainFragment extends Fragment implements BottomBarButtonPath {
   }
 
   public ArrayList<LayoutElementParcelable> addToSmb(
-      SmbFile[] mFile, String path, boolean showHiddenFiles) throws SmbException {
+      @NonNull SmbFile[] mFile, @NonNull String path, boolean showHiddenFiles) throws SmbException {
     ArrayList<LayoutElementParcelable> smbFileList = new ArrayList<>();
+    String extraParams = Uri.parse(path).getQuery();
+
     if (searchHelper.size() > 500) searchHelper.clear();
     for (SmbFile aMFile : mFile) {
       if ((dataUtils.isFileHidden(aMFile.getPath()) || aMFile.isHidden()) && !showHiddenFiles) {
@@ -1430,11 +1427,14 @@ public class MainFragment extends Fragment implements BottomBarButtonPath {
       if (aMFile.isDirectory()) {
         folder_count++;
 
+        Uri.Builder aMFilePathBuilder = Uri.parse(aMFile.getPath()).buildUpon();
+        if (!TextUtils.isEmpty(extraParams)) aMFilePathBuilder.query(extraParams);
+
         LayoutElementParcelable layoutElement =
             new LayoutElementParcelable(
                 getContext(),
                 name,
-                aMFile.getPath(),
+                aMFilePathBuilder.build().toString(),
                 "",
                 "",
                 "",
@@ -1682,7 +1682,7 @@ public class MainFragment extends Fragment implements BottomBarButtonPath {
           }
           */
 
-          s.setStreamSrc(new SmbFile(baseFile.getPath()), baseFile.getSize());
+          s.setStreamSrc(baseFile.getSmbFile(), baseFile.getSize());
           activity.runOnUiThread(
               () -> {
                 try {

--- a/app/src/main/java/com/amaze/filemanager/ui/views/appbar/BottomBar.java
+++ b/app/src/main/java/com/amaze/filemanager/ui/views/appbar/BottomBar.java
@@ -346,10 +346,8 @@ public class BottomBar implements View.OnTouchListener {
 
     switch (openmode) {
       case SFTP:
-        newPath = HybridFile.parseSftpPath(news);
-        break;
       case SMB:
-        newPath = HybridFile.parseSmbPath(news);
+        newPath = HybridFile.parseAndFormatUriForDisplay(news);
         break;
       case OTG:
         newPath = mainActivityHelper.parseOTGPath(news);

--- a/app/src/main/java/com/amaze/filemanager/utils/MainActivityHelper.java
+++ b/app/src/main/java/com/amaze/filemanager/utils/MainActivityHelper.java
@@ -52,6 +52,7 @@ import android.content.BroadcastReceiver;
 import android.content.Context;
 import android.content.Intent;
 import android.content.SharedPreferences;
+import android.net.Uri;
 import android.os.AsyncTask;
 import android.os.Build;
 import android.os.Bundle;
@@ -140,7 +141,15 @@ public class MainActivityHelper {
         "",
         (dialog, which) -> {
           EditText textfield = dialog.getCustomView().findViewById(R.id.singleedittext_input);
-          mkDir(new HybridFile(openMode, path + "/" + textfield.getText().toString()), ma);
+          mkDir(
+              new HybridFile(
+                  openMode,
+                  Uri.parse(path)
+                      .buildUpon()
+                      .appendPath(textfield.getText().toString())
+                      .build()
+                      .toString()),
+              ma);
           dialog.dismiss();
         },
         (text) -> {
@@ -170,7 +179,15 @@ public class MainActivityHelper {
         AppConstants.NEW_FILE_DELIMITER.concat(AppConstants.NEW_FILE_EXTENSION_TXT),
         (dialog, which) -> {
           EditText textfield = dialog.getCustomView().findViewById(R.id.singleedittext_input);
-          mkFile(new HybridFile(openMode, path + "/" + textfield.getText().toString()), ma);
+          mkFile(
+              new HybridFile(
+                  openMode,
+                  Uri.parse(path)
+                      .buildUpon()
+                      .appendPath(textfield.getText().toString())
+                      .build()
+                      .toString()),
+              ma);
           dialog.dismiss();
         },
         (text) -> {

--- a/app/src/main/java/com/amaze/filemanager/utils/SmbUtil.java
+++ b/app/src/main/java/com/amaze/filemanager/utils/SmbUtil.java
@@ -21,12 +21,19 @@
 package com.amaze.filemanager.utils;
 
 import java.io.IOException;
+import java.net.MalformedURLException;
 import java.security.GeneralSecurityException;
 
 import com.amaze.filemanager.filesystem.files.CryptUtil;
+import com.amaze.filemanager.filesystem.smb.CifsContextFactory;
 
 import android.content.Context;
+import android.net.Uri;
 import android.text.TextUtils;
+
+import jcifs.context.SingletonContext;
+import jcifs.smb.NtlmPasswordAuthentication;
+import jcifs.smb.SmbFile;
 
 /**
  * Created by Vishal on 30-05-2017.
@@ -34,6 +41,8 @@ import android.text.TextUtils;
  * <p>Class provides various utility methods for SMB client
  */
 public class SmbUtil {
+
+  public static final String PARAM_DISABLE_IPC_SIGNING_CHECK = "disableIpcSigningCheck";
 
   /** Parse path to decrypt smb password */
   public static String getSmbDecryptedPath(Context context, String path)
@@ -52,7 +61,7 @@ public class SmbUtil {
       String decryptedPassword = CryptUtil.decryptPassword(context, encryptedPassword);
       buffer.append(decryptedPassword);
     }
-    buffer.append(path.substring(path.lastIndexOf("@"), path.length()));
+    buffer.append(path.substring(path.lastIndexOf("@")));
 
     return buffer.toString();
   }
@@ -73,8 +82,19 @@ public class SmbUtil {
       String encryptPassword = CryptUtil.encryptPassword(context, decryptedPassword);
       buffer.append(encryptPassword);
     }
-    buffer.append(path.substring(path.lastIndexOf("@"), path.length()));
+    buffer.append(path.substring(path.lastIndexOf("@")));
 
     return buffer.toString();
+  }
+
+  public static SmbFile create(String path) throws MalformedURLException {
+    Uri uri = Uri.parse(path);
+    boolean disableIpcSigningCheck =
+        Boolean.parseBoolean(uri.getQueryParameter(PARAM_DISABLE_IPC_SIGNING_CHECK));
+    return new SmbFile(
+        path.indexOf('?') < 0 ? path : path.substring(0, path.indexOf('?')),
+        CifsContextFactory.createWithDisableIpcSigningCheck(disableIpcSigningCheck)
+            .withCredentials(
+                new NtlmPasswordAuthentication(SingletonContext.getInstance(), uri.getUserInfo())));
   }
 }

--- a/app/src/main/java/com/amaze/filemanager/utils/SubnetScanner.java
+++ b/app/src/main/java/com/amaze/filemanager/utils/SubnetScanner.java
@@ -21,6 +21,8 @@
 package com.amaze.filemanager.utils;
 
 /** Created by arpitkh996 on 16-01-2016. */
+import static com.amaze.filemanager.filesystem.smb.CifsContextFactory.SMB_URI_PREFIX;
+
 import java.net.UnknownHostException;
 import java.util.ArrayList;
 import java.util.List;
@@ -179,7 +181,7 @@ public class SubnetScanner extends AsyncTask<Void, ComputerParcelable, Void> {
           public void run() {
             for (int i = 0; i < SubnetScanner.RETRY_COUNT; i++) {
               try {
-                SmbFile smbFile = SmbUtil.create("smb://");
+                SmbFile smbFile = SmbUtil.create(SMB_URI_PREFIX);
                 smbFile.setConnectTimeout(5000);
                 SmbFile[] listFiles = smbFile.listFiles();
                 for (SmbFile smbFile2 : listFiles) {

--- a/app/src/main/java/com/amaze/filemanager/utils/SubnetScanner.java
+++ b/app/src/main/java/com/amaze/filemanager/utils/SubnetScanner.java
@@ -179,7 +179,7 @@ public class SubnetScanner extends AsyncTask<Void, ComputerParcelable, Void> {
           public void run() {
             for (int i = 0; i < SubnetScanner.RETRY_COUNT; i++) {
               try {
-                SmbFile smbFile = new SmbFile("smb://");
+                SmbFile smbFile = SmbUtil.create("smb://");
                 smbFile.setConnectTimeout(5000);
                 SmbFile[] listFiles = smbFile.listFiles();
                 for (SmbFile smbFile2 : listFiles) {

--- a/app/src/main/res/layout/smb_dialog.xml
+++ b/app/src/main/res/layout/smb_dialog.xml
@@ -76,12 +76,18 @@
             android:id="@+id/passwordET"
             android:layout_gravity="center_horizontal" />
     </com.google.android.material.textfield.TextInputLayout>
+
     <androidx.appcompat.widget.AppCompatCheckBox
-        android:layout_width="match_parent"
+        android:id="@+id/chkSmbAnonymous"
+        android:layout_width="wrap_content"
         android:layout_height="wrap_content"
-        android:text="@string/anonymous"
-        android:id="@+id/checkBox2"
-        android:layout_gravity="center_horizontal" />
+        android:text="@string/anonymous" />
+    <androidx.appcompat.widget.AppCompatCheckBox
+        android:id="@+id/chkSmbDisableIpcSignature"
+        android:layout_width="wrap_content"
+        android:layout_height="wrap_content"
+        android:text="@string/disableIpcSignature" />
+
     <TextView
         android:layout_width="wrap_content"
         android:id="@+id/wanthelp"

--- a/app/src/main/res/values/strings.xml
+++ b/app/src/main/res/values/strings.xml
@@ -665,5 +665,6 @@
     <string name="move">Move</string>
     <string name="select_save_location">Select location to save</string>
     <string name="scp_default_path">Default Path (Optional)</string>
+    <string name="disableIpcSignature">Disable IPC signing check (For legacy SMBv1 servers - UNSAFE!)</string>
 </resources>
 

--- a/app/src/test/java/com/amaze/filemanager/filesystem/smb/CifsContextFactoryTest.java
+++ b/app/src/test/java/com/amaze/filemanager/filesystem/smb/CifsContextFactoryTest.java
@@ -1,0 +1,89 @@
+/*
+ * Copyright (C) 2014-2020 Arpit Khurana <arpitkh96@gmail.com>, Vishal Nehra <vishalmeham2@gmail.com>,
+ * Emmanuel Messulam<emmanuelbendavid@gmail.com>, Raymond Lai <airwave209gt at gmail.com> and Contributors.
+ *
+ * This file is part of Amaze File Manager.
+ *
+ * Amaze File Manager is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package com.amaze.filemanager.filesystem.smb;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertNotNull;
+import static org.junit.Assert.assertTrue;
+
+import java.util.Properties;
+
+import org.junit.BeforeClass;
+import org.junit.Test;
+
+import androidx.annotation.NonNull;
+
+import jcifs.Config;
+import jcifs.ResolverType;
+import jcifs.context.BaseContext;
+
+public class CifsContextFactoryTest {
+
+  @BeforeClass
+  public static void bootstrap() {
+    Config.registerSmbURLHandler();
+  }
+
+  @Test
+  public void testCreateUsingEmptyPropeties() {
+    BaseContext ctx = CifsContextFactory.create(new Properties());
+    verifyCommonProperties(ctx);
+  }
+
+  @Test
+  public void testCreateWithNull() {
+    BaseContext ctx = CifsContextFactory.create(null);
+    verifyCommonProperties(ctx);
+  }
+
+  @Test
+  public void testCreateUsingCustomProperties() {
+    Properties p = new Properties();
+    p.setProperty("jcifs.smb.client.ipcSigningEnforced", "false");
+    BaseContext ctx = CifsContextFactory.create(p);
+    verifyCommonProperties(ctx);
+    assertFalse(ctx.getConfig().isIpcSigningEnforced());
+  }
+
+  @Test
+  public void testRepeatedCreateWithCustomProperties() {
+    Properties p = new Properties();
+    p.setProperty("jcifs.smb.client.ipcSigningEnforced", "false");
+    BaseContext ctx = CifsContextFactory.create(p);
+    verifyCommonProperties(ctx);
+    assertFalse(ctx.getConfig().isIpcSigningEnforced());
+
+    p = new Properties();
+    p.setProperty("jcifs.smb.client.ipcSigningEnforced", "true");
+    ctx = CifsContextFactory.create(p);
+    verifyCommonProperties(ctx);
+    assertTrue(ctx.getConfig().isIpcSigningEnforced());
+  }
+
+  private void verifyCommonProperties(@NonNull BaseContext ctx) {
+    assertNotNull(ctx);
+    assertEquals(ResolverType.RESOLVER_BCAST, ctx.getConfig().getResolveOrder().get(0));
+    assertEquals(-1 * 60, ctx.getConfig().getNetbiosCachePolicy());
+    assertEquals(30000, ctx.getConfig().getResponseTimeout());
+    assertEquals(5000, ctx.getConfig().getNetbiosRetryTimeout());
+  }
+}

--- a/app/src/test/java/com/amaze/filemanager/filesystem/ssh/test/TestUtils.java
+++ b/app/src/test/java/com/amaze/filemanager/filesystem/ssh/test/TestUtils.java
@@ -20,6 +20,8 @@
 
 package com.amaze.filemanager.filesystem.ssh.test;
 
+import static com.amaze.filemanager.filesystem.ssh.SshConnectionPool.SSH_URI_PREFIX;
+
 import java.io.IOException;
 import java.io.StringWriter;
 import java.security.KeyPair;
@@ -77,7 +79,7 @@ public abstract class TestUtils {
       privateKeyContents = writer.toString();
     }
 
-    StringBuilder fullUri = new StringBuilder().append("ssh://").append(validUsername);
+    StringBuilder fullUri = new StringBuilder().append(SSH_URI_PREFIX).append(validUsername);
 
     if (validPassword != null) fullUri.append(':').append(validPassword);
 

